### DISCUSSION TITLE
Timestamp format in JSON logging not conforming to timestamp pattern

### DISCUSSION THREAD

**Discussion by persona94** [odygrd/quill#601] [General]
I used the code from the JSON logging page

```
#include "quill/Backend.h"
#include "quill/Frontend.h"
#include "quill/LogMacros.h"
#include "quill/Logger.h"
#include "quill/sinks/JsonConsoleSink.h"
#include <string>

int main()
{
    quill::BackendOptions backend_options;
    quill::Backend::start(backend_options);

    auto json_sink = quill::Frontend::create_or_get_sink<quill::JsonConsoleSink>("json_sink_1");

    // When logging json, it is ideal to set the logging pattern to empty to avoid unnecessary message formatting.
    quill::Logger* logger = quill::Frontend::create_or_get_logger(
      "json_logger", std::move(json_sink), quill::PatternFormatterOptions { "", "%H:%M:%S.%Qns", quill::Timezone::GmtTime });

    int var_a = 123;
    std::string var_b = "test";

    // Log via the convenient LOGJ_ macros
    LOGJ_INFO(logger, "A json message", var_a, var_b);

    // Or manually specify the desired names of each variable
    LOG_INFO(logger, "A json message with {var_1} and {var_2}", var_a, var_b);
}
```

The output is

```
{"timestamp":"1728071935315527792","file_name":"main.cpp","line":"23","thread_id":"1412467","logger":"json_logger","log_level":"INFO","message":"A json message {var_a}, {var_b}","var_a":"123","var_b":"test"}
{"timestamp":"1728071935315569792","file_name":"main.cpp","line":"26","thread_id":"1412467","logger":"json_logger","log_level":"INFO","message":"A json message with {var_1} and {var_2}","var_1":"123","var_2":"test"}
```

Why are the timestamps in epoch and not human readable format?

**Comment by odygrd** [comment]
Currently, you can have both a normal output and a JSON output simultaneously. However, the `JsonSink` is performing its own formatting and does not follow the `PatternFormatter` used for regular files.

The `JsonSink` handles formatting here:
- [JsonFileSink](https://github.com/odygrd/quill/blob/bb7bbfa8805cc6c79beb07e0836a459d0928dbe7/include/quill/sinks/JsonFileSink.h#L92)
- [JsonConsoleSink](https://github.com/odygrd/quill/blob/bb7bbfa8805cc6c79beb07e0836a459d0928dbe7/include/quill/sinks/JsonConsoleSink.h#L76)

I recommend creating your own `JsonSink` type by deriving from these classes and implementing your own `write_log` function with the desired format. You can refer to this example for guidance:
- [User-Defined Sink Example](https://github.com/odygrd/quill/blob/master/examples/user_defined_sink.cpp)

I will look into how this can be improved in future versions.

**Reply by persona94** [reply]
Ah gotcha. Thank you for the very prompt response! I'll look into a user-defined sink because I need to embed some thread specific data anyway
